### ISSUE TITLE
Bastion and usage with non-root user for installation

### ISSUE THREAD

**Issue by torwen1** [ocp-power-automation/ocp4-upi-powervm#183]
Hello team,

During some tests we found some issues in situations, where customers doesn't allow direct root access to the bastion host. These customers having a normal user, which has the sudo permission to execute all commands as root (e.g. sudo systemctl restart service).

I'm not sure, if what we have discovered is a feature request, or a problem.

Here the summary what we have discovered:

In the var.tfvars we configured the rhel_user to customers non-root user for installation.
Most/all of the system related commands in your TF modules are already using sudo, but we also had to ensure, that the Ansible playbooks of ocp4-helpernode and ocp4-playbook are executed as root. So, we configured the ansible_extra_options      = "-v --become" to ensure this.

Now comes the trouble we run into:
All file uploads from the TF modules are placed in the HOME of the specified RHEL user, but the Ansible playbooks are placing files in the HOME of root (because of the --become parameter). This ends in the situation, that the Ansible playbooks are failing at different stages because of the mix where files are placed into etc.

To fully support installations without direct root access and with sudo, we would need an option in the var.tfvars, where we are able to define a common work directory for TF modules, and all the Ansible playbooks.

We also discovered that some Ansible playbooks using workdir variables and sometimes they have a hard-coded ~ instead. I guess that this needs to be reported against the different Ansible playbook projects, instead of here?

**Comment by yussufsh** [comment]
Thanks for reporting this.
This is an important requirement and I will say got less importance as we kept adding new features to the code. I agree that most of our installations were managed by root user on bastion but having a sudo user is a valid usecase that we need to work on with helpernode & install playbooks.

I will take some time and analyse the areas where the updates are needed to fix this.

**Comment by torwen1** [comment]
I'm happy to provide you a diff with what we needed to change in all the different files to get it running. I think this should give you an idea of what is affected.

**Comment by torwen1** [comment]
I did this kind of change for a customer and faced some additional parts, which are affected by having a non-root access option for the bastion node:
1) Ansible playbooks are need to run with the --become option
I think this one is clear, why it is needed.
2) Because Ansible playbooks have to run with --become, the usage of the $HOME directory is creating a mess. Terraform will upload the playbooks to the $HOME of the non-root user, but the playbooks generated files are placed in the $HOME of the root user. This required an option to define a central working directory. At the customer I solved it by adding a new var.tfvars variable. I changed the tf files, but also had to pass that to both Ansible playbooks ocp4-helpernode and ocp4-playbooks. If I remember well, ocp4-helpernode did not needed a code change, but ocp4-playbooks.
3) There is an issue with the SSH public key. During the deployment of the bastion node, the public key gets added to the authorised_keys file of the root user. But we would try to login with non-root, which will fail because the key is not in this users authorised_key file. At the mentioned customer implementation I solved it by having an install SSH key in the image, which gets exchanged as soon as TF can reach the bastion by SSH. For me it is more a workaround, then a solution. I would like to see, that I can pass this somehow as config-parameter to cloud-init. An approach would be to use the "users" part of cloud-init https://cloudinit.readthedocs.io/en/latest/topics/examples.html. Just have to find a way to pass this as cloud-init config parameters during deployment request (any idea is highly appreciated). The rest of the code changes is already available and in use by a customer.
